File: Zotlabs/Lib/JSalmon.php
<?php

namespace Zotlabs\Lib;

use Zotlabs\Web\HTTPSig;

class JSalmon
{

    public static function sign($data, $key_id, $key, $data_type = 'application/x-zot+json')
    {

        $data = base64url_encode(json_encode($data, true), true); // strip padding
        $encoding = 'base64url';
        $algorithm = 'RSA-SHA256';

        $data = preg_replace('/\s+/', '', $data);

        // precomputed base64url encoding of data_type, encoding, algorithm concatenated with periods

        $precomputed = '.' . base64url_encode($data_type, true) . '.YmFzZTY0dXJs.UlNBLVNIQTI1Ng';

        $signature = base64url_encode(Crypto::sign($data . $precomputed, $key), true);

        return ([
            'signed' => true,
            'data' => $data,
            'data_type' => $data_type,
            'encoding' => $encoding,
            'alg' => $algorithm,
            'sigs' => [
                'value' => $signature,
                'key_id' => base64url_encode($key_id, true)
            ]
        ]);

    }

    public static function verify($x)
    {

        logger('verify');
        $ret = ['results' => []];

        if (!is_array($x)) {
            return false;
        }
        if (!(array_key_exists('signed', $x) && $x['signed'])) {
            return false;
        }

        $signed_data = preg_replace('/\s+/', '', $x['data']) . '.'
            . base64url_encode($x['data_type'], true) . '.'
            . base64url_encode($x['encoding'], true) . '.'
            . base64url_encode($x['alg'], true);

        $key = HTTPSig::get_key(EMPTY_STR, 'zot6', base64url_decode($x['sigs']['key_id']));
        logger('key: ' . print_r($key, true));
        if ($key['portable_id'] && $key['public_key']) {
            if (Crypto::verify($signed_data, base64url_decode($x['sigs']['value']), $key['public_key'])) {
                logger('verified');
                $ret = ['success' => true, 'signer' => $key['portable_id'], 'hubloc' => $key['hubloc']];
            }
        }

        return $ret;

    }

    public static function unpack($data)
    {
        return json_decode(base64url_decode($data), true);
    }


}

File: util/py/jsalmon.py


from libzot import *
import re
import json


class JSalmon:

    def sign(data,key_id,key,data_type = 'application/x-nomad+json'):
        data = base64urlnopad_encode(data.encode("utf-8"))
        encoding = 'base64url'
        algorithm = 'RSA-SHA256'

        data = re.sub(r'\s+',"",data)
        fields = data + "." + base64urlnopad_encode(data_type.encode("utf-8")) + "." + base64urlnopad_encode(encoding.encode("utf-8")) + "." + base64urlnopad_encode(algorithm.encode("utf-8"))
        signature = base64urlnopad_encode(rsa_sign(fields,key).encode("utf-8"))
        return {
            'signed' : True,
            'data' : data,
            'data_type' : data_type,
            'encoding' : encoding,
            'alg'  : algorithm,
            'sigs' : { 'value' : signature, 'key_id' : base64urlnopad_encode(key_id.encode("utf-8")) }}


    def verify(x,key):
        if x['signed'] != True:
            return False

        signed_data = re.sub(r'\s+','', x['data'] + "." + base64urlnopad_encode(x['data_type'].encode("utf-8")) + "." + base64urlnopad_encode(x['encoding'].encode("utf-8")) + "." + base64urlnopad_encode(x['alg'].encode("utf-8")))

        binsig = base64urlnopad_decode(x['sigs']['value'])
        
        if rsa_verify(signed_data,binsig,key) == True:
            return True

        return False

    def unpack(data):
        return json.loads(base64urlnopad_decode(data))


        
#if __name__=="__main__":
#    prvkey,pubkey = generate_rsa_keypair()

#    s = JSalmon.sign('abc123','mykeyid',prvkey)
#    print (s)

#    if JSalmon.verify(s,pubkey):
#        print ('verified')
#    else:
#        print ('failed')
